### PR TITLE
examples, bug fixes, optimizations

### PR DESCRIPTION
**Pull request by dannyech**
1. Added analyzer type checking examples 
2. Fixed three bugs in Analyzer that were preventing merge sort from working. The bugs centered around the 'head()` and `cons()` functions that we hacked into the analyzer
3. only importing matplotlib when necessary so our execution is faster

### CODE REVIEW

**Comment by dannyech** [comment]
for review @akalicki @alexandramedway @lillyfwang

File: docs/examples/checking_and_inference/arith_fail.ods [hunk 1]
```diff
@@ -0,0 +1,3 @@
+do n = 2
+
+do fail = !n /* !n invalid b/c n as type NUM */
```
Comment (akalicki) [inline docs/examples/checking_and_inference/arith_fail.ods]: nit: newline at end

File: docs/examples/checking_and_inference/call.ods [hunk 2]
```diff
@@ -0,0 +1,3 @@
+do append_hello = (l) -> return "hello" :: l
+
+do success = append_hello(["world"])
```
Comment (akalicki) [inline docs/examples/checking_and_inference/call.ods]: nit: newline at end

File: docs/examples/checking_and_inference/call_fail.ods [hunk 3]
```diff
@@ -0,0 +1,4 @@
+do append_hello = (l) -> return "hello" :: l
+
+/* fail because append_hello expects List of String, not List of Num */
+do fail = append_hello([42])
```
Comment (akalicki) [inline docs/examples/checking_and_inference/call_fail.ods]: nit: newline at end

File: docs/examples/checking_and_inference/order.txt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+1. arith
```
Comment (akalicki) [inline docs/examples/checking_and_inference/order.txt]: what does this file mean?

File: docs/examples/checking_and_inference/params.ods [hunk 5]
```diff
@@ -0,0 +1,4 @@
+do and = (x, y) ->
+    do result = x && y
+    do success = x || y
+    return result
```
Comment (akalicki) [inline docs/examples/checking_and_inference/params.ods]: nit: newline at end

File: docs/examples/checking_and_inference/params_as_functions.ods [hunk 6]
```diff
@@ -0,0 +1,14 @@
+do call = (f, x, y, dummy) ->
+    
+    /* Analyzer now knows f is function that takes two paramters. The first
+     * must be a num, but the second is unkown because the paramter x still has
```
Comment (akalicki) [inline docs/examples/checking_and_inference/params_as_functions.ods]: sp: unknown

File: docs/examples/checking_and_inference/params_as_functions.ods [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+     * type unkown */
```
Comment (akalicki) [inline docs/examples/checking_and_inference/params_as_functions.ods]: here too

File: docs/examples/checking_and_inference/params_as_functions.ods [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+    do f(2, x)
+
+    /* Analyzer now knows x is a Num and that the second argument of f is a Num */
+    do f(x, 2)
+
+    /* y must now be a Num and f must now return a Bool */
+    return f(x, y) && f(y, x)
+
+/* dummy can be any type because it is never used within the function */
```
Comment (akalicki) [inline docs/examples/checking_and_inference/params_as_functions.ods]: nit: newline at end

File: docs/examples/checking_and_inference/params_fail.ods [hunk 9]
```diff
@@ -0,0 +1,5 @@
+do and = (x, y) ->
+    do result = x && y
+    /*invalid because x and y have type Bool, but addition expects Num */
```
Comment (akalicki) [inline docs/examples/checking_and_inference/params_fail.ods]: nit: space after first comment

File: docs/examples/checking_and_inference/params_fail.ods [hunk 10]
```diff
@@ -0,0 +1,5 @@
...
+    do fail = x + y
+    return result
```
Comment (akalicki) [inline docs/examples/checking_and_inference/params_fail.ods]: nit: newline at end

File: docs/examples/checking_and_inference/recursion.ods [hunk 11]
```diff
@@ -0,0 +1,2 @@
+do fib = (n) ->
+    return if n <= 1 then n else fib(n - 1) + fib(n - 2)
```
Comment (akalicki) [inline docs/examples/checking_and_inference/recursion.ods]: nit: newline at end

File: docs/examples/checking_and_inference/recursion2_fail.ods [hunk 12]
```diff
@@ -0,0 +1,7 @@
+do demonstrate_constraining = (x) ->
+    do demonstrate_constraining(2) /* constrains x to Num here */
+    
+    /* will fail below because x is constrained to Num above and cannot be a 
+     * list of lists of strings */
+    do fail = ["I'm going to brak things"] :: x
+    return "wow!"
```
Comment (akalicki) [inline docs/examples/checking_and_inference/recursion2_fail.ods]: see above

File: docs/examples/checking_and_inference/recursion_fail.ods [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+       inf_recursion() + 2
+    b) However, the return statement returns true, which is a Bool. The
+       Analyzer recognizes this error and fails
+
+    The function below, however, would pass:
+    
+        do inf_recursion = () ->
+            do inf_recursion() + 2
+            return 2
+*/
```
Comment (akalicki) [inline docs/examples/checking_and_inference/recursion_fail.ods]: see above

File: compiler/utils.ml [hunk 14]
```diff
@@ -18,4 +18,4 @@ let str_of_file file =
     | Some(s) -> aux (s :: acc) in
   aux []
 
-let conclude_program () = "if PLOT:\n\tplt.show()"
+let conclude_program () = "if PLOT:\n\timport matplotlib.pyplot as plt\n\tplt.show()"
```
Comment (akalicki) [inline compiler/utils.ml]: is it necessary to import again here? won't plt.show() only be called if the function has been run and plt already imported? or is plt local to that function?

**Comment by akalicki** [comment]
mostly style nits, see above

File: docs/examples/checking_and_inference/order.txt (hunk 4 above)
Comment (dannyech) [inline docs/examples/checking_and_inference/order.txt]: it's the order in which i would show Edwards the examples - this is just a rough draft, liable to change + input from you guys. It sounds like I may have made too many examples, and may need to condense into only one or two.

@akalicki

File: compiler/utils.ml (hunk 14 above)
Comment (dannyech) [inline compiler/utils.ml]: @akalicki it's only imported local to that function - there might be a way around this, but i am unaware

**Pull request merged**